### PR TITLE
Fix TransposeSinking transformation for Gather op

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
Fix: restore  the graph state (delete previously inserted squeeze op) before `return false` .
Added additional checks, rt_info and names copying and test.

### Tickets:
 - *110588*

### CODE REVIEW

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp [hunk 1]
```diff
@@ -169,30 +169,37 @@ TSGatherBackward::TSGatherBackward() {
...
-            if (success) {
-                size_t j = 0;
-                for (size_t i = 0; i < shape.size(); ++i) {
-                    if (shape[i] != new_shape[j] && shape[i] == 1) {
-                        axes_val.push_back(i);
-                        continue;
-                    } else if (shape[i] != new_shape[j]) {
+            if (new_out_pshape.is_static()) {
+                auto shape = out_pshape.get_shape();
+                auto new_shape = new_out_pshape.get_shape();
```
Comment (t-jankowski) [inline src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp:184]:
```suggestion
                const auto shape = out_pshape.get_shape();
                const auto new_shape = new_out_pshape.get_shape();
```

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp [hunk 2]
```diff
@@ -169,30 +169,37 @@ TSGatherBackward::TSGatherBackward() {
...
+                success = shape != new_shape;
+                if (success) {
+                    size_t j = 0;
+                    for (size_t i = 0; i < shape.size(); ++i) {
+                        if (shape[i] != new_shape[j] && shape[i] == 1) {
+                            axes_val.push_back(i);
+                            continue;
+                        } else if (shape[i] != new_shape[j]) {
+                            success = false;
```
Comment (rkazants) [inline src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp:193]: should not we break `for` here?

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp [hunk 3]
```diff
@@ -230,6 +240,11 @@ TSGatherBackward::TSGatherBackward() {
             for (const auto& input : target_inputs) {
                 input.replace_source_output(unsqueeze);
             }
+            unsqueeze->output(0).get_tensor_ptr()->add_names(main_node->output(0).get_names());
```
[the remark this answers is not in the document]
Comment (rkazants) [reply src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp:243]: Isn't it the same?

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp [hunk 4]
```diff
@@ -169,30 +169,37 @@ TSGatherBackward::TSGatherBackward() {
...
-                    if (shape[i] != new_shape[j] && shape[i] == 1) {
-                        axes_val.push_back(i);
-                        continue;
-                    } else if (shape[i] != new_shape[j]) {
+            if (new_out_pshape.is_static()) {
+                auto shape = out_pshape.get_shape();
+                auto new_shape = new_out_pshape.get_shape();
+                success = shape != new_shape;
+                if (success) {
+                    size_t j = 0;
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp:204]: added an ascii graph

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp (hunk 2 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp:210]: fixed

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp [hunk 5]
```diff
@@ -169,30 +169,37 @@ TSGatherBackward::TSGatherBackward() {
...
                         success = false;
                     }
-                    j++;
-                }
-                if (j != new_shape.size()) {
-                    success = false;
                 }
             }
-            if (!success) {
+            if (!success && squeeze) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp]: fixed

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp]: fixed

File: src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp (hunk 1 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/transpose_sinking/ts_gather.cpp]: fixed

**Review by t-jankowski — APPROVED** [review]

**Review by rkazants — APPROVED** [review]

**Pull request merged**